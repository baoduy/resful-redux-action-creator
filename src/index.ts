/**
 * The method to create Reducer from Actions
 */
export { createReducer } from './reduxReducerCreator';

/**
 * The method to create Redux Actions from Api Actions crated by 'restful-action-creator'
 */
export { createActions } from './reduxActionCreator';

/**
 * The state helper methods. allow to merge Data from Payload to State.
 * This method design to work with 'welldone-software/redux-toolbelt'
 */
export { mergeData, removeItems } from './reduxStateHelper';
