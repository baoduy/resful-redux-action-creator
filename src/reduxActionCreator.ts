import {
  ActionOptions,
  ReduxActionCollection,
  RestActionCollection
} from './reduxDefinition';

import createReduxTool from './reduxTool';

const ignoreExtraParam = <TAsyncFunc extends Function>(
  asyncFunc: TAsyncFunc
) => (...args: Array<any>) => {
  if (args.length > 0) args[args.length - 1] = undefined;
  return asyncFunc(...args);
};
/** Api actions had been created by restful-action-creator*/
export const createActions = <TActions extends RestActionCollection>(
  restActions: TActions,
  options: ActionOptions = {}
): ReduxActionCollection<TActions> => {
  const name = restActions.name || options.prefix;
  if (!name) throw 'prefix and Api.name are empty.';

  const tool = createReduxTool(`${name.toUpperCase()}@`);
  const actions = { original: restActions };

  //Add action from Api.
  Object.keys(restActions).forEach(key => {
    const action = restActions[key];
    if (typeof action !== 'function') return;

    actions[key] = tool.createAsyncAction(
      key.toUpperCase(),
      options.ignoreExtraParam ? ignoreExtraParam(action) : action,
      {}
    );
  });

  return <ReduxActionCollection<TActions>>actions;
};
