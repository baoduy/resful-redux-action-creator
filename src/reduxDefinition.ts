import { Action, AnyAction, Dispatch, Reducer } from 'redux';
import { AxiosPromise, AxiosResponse } from 'axios';

import { AsyncActionCreator } from 'redux-toolbelt';

export type DataGetterFunc = (state: any, action: any) => any;
export type MetaGetterFunc = (api: Function) => any;

export type DispatchableFunc<T = any> = (
  dispatch: Dispatch<AnyAction>
) => Promise<T> | AxiosResponse<T>;

type RestActionWithParams<T = any> = (
  ...params: Array<any>
) => AxiosPromise<T> | Promise<T>;

type RestActionWithoutParams = () => AxiosPromise<void> | Promise<void>;

export type RestAction<T = any> =
  | RestActionWithParams<T>
  | RestActionWithoutParams;

export interface Item {
  id?: string | number;
  name?: string | number;
  [key: string]: any;
}

export type MetaDataItem<T = any> = { items?: Array<T>; [key: string]: any };
export type DataItem<T = any> = Array<T> | MetaDataItem<T>;

// Names of properties in T with types that include undefined
// type OptionalPropertyNames<T> = {
//   [K in keyof T]: undefined extends T[K] ? K : never
// }[keyof T];

// Common properties from L and R with undefined in R[K] replaced by type in L[K]
// type SpreadProperties<L, R, K extends keyof L & keyof R> = {
//   [P in K]: L[P] | Exclude<R[P], undefined>
// };

// type Props<T> = { [K in keyof T]: T[K] };

// export type Spread<L, R> = Props<
//   Pick<L, Exclude<keyof L, keyof R>> &
//     Pick<R, Exclude<keyof R, OptionalPropertyNames<R>>> &
//     Pick<R, Exclude<OptionalPropertyNames<R>, keyof L>> &
//     SpreadProperties<L, R, OptionalPropertyNames<R> & keyof L>
// >;

export interface RestActionCollection<T extends Action = any> {
  name: string | undefined;
  [key: string]: string | AsyncActionCreator<T> | undefined;
}

export interface ReduxAction<T = any> extends AnyAction {
  (...params: Array<any>): DispatchableFunc<T>;
  success: (...params: Array<any>) => RestAction<T>;
  failure: (...params: Array<any>) => RestAction<T>;
  progress: (...params: Array<any>) => RestAction<T>;
  cancel: (...params: Array<any>) => RestAction<T>;
}

export type ReduxActionCollection<
  TActions extends RestActionCollection<any>
> = { [K in keyof TActions]: AsyncActionCreator<any> } & {
  /** The original RestActionCollection */
  original: TActions;
};

export interface ReducerFunc<S = any, A extends ReduxAction = any>
  extends Reducer<S, A> {
  (state: S, action: A): S;
  reducers: Array<ReduxAction>;
  actionName: string;
}

export interface dataGetterOption {
  [key: string]: DataGetterFunc | false;
}

export type ReducerOptions = {
  name?: string;
  meta?: MetaGetterFunc;
  defaultData: any;
  shouldSpread?: boolean;
  /** Set of dataGetters for each actions {[action]:dataGetter,...}.
   If dataGetter for a action is not found the defaultDataGetter will be picked up */
  dataGetters?: dataGetterOption /** default DataGetter for actions */;
  defaultDataGetter?: DataGetterFunc;
  dataProp?: string;
  shouldDestroyData?: boolean;
  shouldDestroyDataOnError?: boolean;
  shouldSetError?: boolean;
  shouldSetData?: boolean;
};

export interface ActionOptions {
  prefix?: string;
  /** Ignore the extract parameter of redux belt {getState, dispatch, extraThunkArg}.
   * The last parameter will be drop before pass to asyncFunc
   */

  ignoreExtraParam?: boolean;
}

export type IdSelectorFunc = <T>(item: T) => any;

export type ValueSelector = <T>(
  prop: string,
  originalValue: T,
  latestValue: T
) => any;

export type DataPropSelector = <T>(data: T) => any;

export interface MergeDataOptions {
  /** value transformer allows to convert, format and apply the calculation for final value from original value and latest value which can be identified via prop name.
   * this will be used for updateObjectProperties execution
   */
  valueSelector?: ValueSelector;
  /** Provide the Id selector if array doesn't have id field.
   * this use be used for upsertItemsById execution.
   */
  idSelector?: IdSelectorFunc;
}

export interface MergeStateOptions extends MergeDataOptions {
  /** provide a Data prop selector and transform for action.playload.data */
  actionDataPropSelector?: DataPropSelector;
}
