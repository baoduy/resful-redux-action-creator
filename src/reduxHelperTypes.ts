import { AnyAction, Dispatch, Reducer } from 'redux';
import { AxiosPromise, AxiosResponse } from 'axios';

export type DataGetterFunc = (state: any, action: any) => any;
export type MetaGetterFunc = (api: Function) => any;
export type DispatchableFunc<T> = (dispatch: Dispatch<AnyAction>) => Promise<T>;

export type NamedObj = { name: string };

export type RestAction<T = any> = (
  ...params: Array<any>
) => AxiosPromise<T> | Promise<T>;

export interface RestActionCollection extends NamedObj {
  [key: string]: string | RestAction | undefined;
}

export interface ReduxAction<T = any> extends AnyAction {
  (...params: Array<any>):
    | DispatchableFunc<T>
    | DispatchableFunc<AxiosResponse<T>>;
  success: (...params: Array<any>) => RestAction<T>;
  failure: (...params: Array<any>) => RestAction<T>;
  progress: (...params: Array<any>) => RestAction<T>;
  cancel: (...params: Array<any>) => RestAction<T>;
}

export type ReduxActionCollection<TActions extends RestActionCollection> = {
  [K in keyof TActions]: string | ReduxAction | undefined
};

export interface ReducerFunc<S = any, A extends ReduxAction = any>
  extends Reducer<S, A> {
  (state: S, action: A): S;
  reducers: Array<ReduxAction>;
  actionName: string;
}

export interface dataGetterOption {
  [key: string]: DataGetterFunc | false;
}

export type ReducerOptions = {
  name?: string;
  meta?: MetaGetterFunc;
  defaultData: any;
  shouldSpread?: boolean;
  /** Set of dataGetters for each actions {[action]:dataGetter,...}.
   If dataGetter for a action is not found the defaultDataGetter will be picked up */
  dataGetters?: dataGetterOption /** default DataGetter for actions */;
  defaultDataGetter?: DataGetterFunc;
  dataProp?: string;
  shouldDestroyData?: boolean;
  shouldDestroyDataOnError?: boolean;
  shouldSetError?: boolean;
  shouldSetData?: boolean;
};

export interface ActionOptions {
  prefix?: string;
}

export type IdSelectorFunc = (item: any) => any;

type ValueSelector = (
  prop: string,
  originalValue: any,
  latestValue: any
) => any;
type DataPropSelector = (data: any) => any;

export interface MergeDataOptions {
  /** value transformer allows to convert, format and apply the calculation for final value from original value and latest value which can be identified via prop name.
   * this will be used for updateObjectProperties execution
   */
  valueSelector?: ValueSelector;
  /** Provide the Id selector if array doesn't have id field.
   * this use be used for upsertItemsById execution.
   */
  idSelector?: IdSelectorFunc;
}
export interface MergeStateOptions extends MergeDataOptions {
  /** provide a Data prop selector and transform for action.playload.data */
  actionDataPropSelector?: DataPropSelector;
}
