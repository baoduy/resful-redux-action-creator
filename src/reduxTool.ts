import composeReducers from 'redux-toolbelt/lib/composeReducers';
import makeAsyncReducer from 'redux-toolbelt/lib/makeAsyncReducer';
import { makeThunkAsyncActionCreator } from 'redux-toolbelt-thunk';

//import makeActionCreator from 'redux-toolbelt/lib/makeActionCreator';
//import makeReducer from 'redux-toolbelt/lib/makeReducer';
const argsMapper = (payload: any, meta: any) => ({ payload, meta });

/** The action creator for both async and non async action */
export default (prefix: string) => ({
  /** Create Async action */
  createAsyncAction: makeThunkAsyncActionCreator.withDefaults({
    prefix,
    argsMapper
  }),
  /** Create non async action */
  // createAction: makeActionCreator.withDefaults({
  //   prefix
  // }),
  /** Create reducer for async actions */
  createAsyncReducer: makeAsyncReducer,
  /** Create reducer for non async actions */
  //createReducer: makeReducer,
  /** Create combined reducers to single reducer */
  createComposeReducer: composeReducers
});
