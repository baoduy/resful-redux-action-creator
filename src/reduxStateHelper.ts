import {
  upsertItemsById,
  updateItemsById,
  updateObjectProperties,
  removeItemsById
} from 'redux-toolbelt-immutable-helpers';
import { MergeDataOptions, IdSelectorFunc } from './reduxHelperTypes';

export interface Item {
  id: string | number;
}

/**
 * Insert or Update array from array based on idSelector,
 * @param original The original array
 * @param latest The latest array contains list items which will be update or insert to the original array.
 * @param idSelector default is item => item.id
 */
const upsertItems = <T extends Item>(
  original: Array<T>,
  latest: Array<T>,
  idSelector?: IdSelectorFunc
): Array<T> => {
  const final = upsertItemsById(original, latest, idSelector);
  return updateItemsById(final, latest, idSelector);
};

/** The helper to mege data props from payload to redux state. The data props is defined by redix-toolbelt.
 * Array properties will be merge based on 'id' field. if don't have id file the provide the id selector.
 */
export const mergeData = (
  original: any,
  latest: any,
  options: MergeDataOptions = { idSelector: i => i.id }
) => {
  if (!latest) return original;
  if (!original) return latest;

  if (Array.isArray(latest) || Array.isArray(original))
    return upsertItems(original, latest, options.idSelector);

  let arrayObj = {};
  let propsTobeUpdate = new Array<string>();

  //Merge Array Based Id
  Object.keys(latest).forEach(k => {
    if (!Array.isArray(latest[k])) {
      propsTobeUpdate.push(k);
      return;
    }

    const lArray = latest[k];
    const oArray = original[k];

    arrayObj[k] = upsertItems(oArray, lArray, options.idSelector);
  });

  //update properties
  const lasteData = updateObjectProperties(
    original,
    propsTobeUpdate,
    (value: any, prop: string) => {
      if (!options.valueSelector) return latest[prop];
      return options.valueSelector(prop, value, latest[prop]);
    }
  );

  return updateObjectProperties(lasteData, arrayObj);
};

/**
 * Remove items from array based on Id field
 * @param original The original array
 * @param id the list of id or just an Id will be remove from original array.
 * @param idSelector default is item => item.id
 */
export const removeItems = <T extends Item>(
  original: Array<T>,
  id: Item | Array<Item>,
  idSelector?: IdSelectorFunc
) =>
  original
    ? removeItemsById(original, Array.isArray(id) ? id : [id], idSelector)
    : [];

// export const mergeState = (
//   state: any,
//   action: any,
//   options: MergeStateOptions = {}
// ) => {
//   const { data, ...others } = action.payload;

//   const newData = mergeData(
//     state.data,
//     options.actionDataPropSelector
//       ? options.actionDataPropSelector(data)
//       : data,
//     options
//   );
//   const newState = updateObjectProperties(state, others);
//   newState.data = newData;
//   return newState;
// };
