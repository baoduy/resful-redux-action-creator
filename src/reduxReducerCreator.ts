import {
  DataGetterFunc,
  ReducerOptions,
  ReduxActionCollection,
  RestActionCollection
} from './reduxDefinition';

import { Reducer } from 'redux';
import createReduxTool from './reduxTool';
import { mergeData } from './reduxStateHelper';

const ignoreDataGetter: DataGetterFunc = s => s;

const defaultOptions: ReducerOptions = {
  defaultData: {},
  shouldSetError: false,
  shouldSetData: true,
  shouldDestroyData: false,
  shouldDestroyDataOnError: false,
  defaultDataGetter: (s, a) => mergeData(s.data, a.payload.data || a.payload)
};

export const createReducer = <TActions extends RestActionCollection>(
  actions: ReduxActionCollection<TActions>,
  options: ReducerOptions = defaultOptions
) => {
  options = Object.assign({}, defaultOptions, options);
  const name = actions.original.name || options.name;
  if (!name) throw 'name of ReduxActionCollection<TActions> is required';

  const tool = createReduxTool(name);

  const reducers = new Array<Reducer>();

  Object.keys(actions).forEach(key => {
    const ac = actions[key];
    if (typeof ac !== 'function') return;

    let dataGetter: DataGetterFunc | undefined = undefined;

    if (options.dataGetters) {
      //If an Action in dataGettes is false => don't update the Redux store value.
      //The Reducer just return the store back.
      if (options.dataGetters[ac as any] === false)
        dataGetter = ignoreDataGetter;
      //Get the Reducer from dataGetters
      else dataGetter = options.dataGetters[ac as any] as DataGetterFunc;
    }

    //If there is no Reducer found in dataGetters the use defaultDataGetter
    if (!dataGetter) dataGetter = options.defaultDataGetter;

    //There is no Reducer found for action.
    if (!dataGetter)
      throw `There is no dataGetter found for ${key} in ${actions.name}`;

    const reducer = tool.createAsyncReducer(ac, {
      ...options,
      dataGetter: dataGetter
    });

    //Add reducer to collection
    reducers.push(reducer);
  });

  const composeReducer = tool.createComposeReducer(...reducers);
  //(composeReducer as any).actionName = actions.name;
  //(composeReducer as any).reducers = reducers;

  return composeReducer;
};
