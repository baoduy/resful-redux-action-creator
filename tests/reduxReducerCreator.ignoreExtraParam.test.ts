import { createActions, createReducer } from '../src';

import postApi from './api/postApi';

describe('Test Reducer Creator', () => {
  test('Test Create Redux Real Get All', async () => {
    const defaultDataGetter = (_s: any, { payload }: any) => payload;
    const dataGetter = (_s: any, { payload }: any) => payload;

    const actions = createActions(postApi, { ignoreExtraParam: true });
    const reducer = createReducer(actions, {
      defaultData: {},
      dataGetters: {
        [actions.get as any]: dataGetter
      },
      defaultDataGetter
    });

    if (typeof actions.get === 'function') {
      const result = await actions.get()(jest.fn());

      const rs = reducer({}, actions.get.success(result.data));
      expect(rs).toMatchSnapshot();
    }
  });

  test('Test Create Redux Real Reducer', async () => {
    const defaultDataGetter = (_s: any, { payload }: any) => payload;
    const dataGetter = (_s: any, { payload }: any) => payload;

    const actions = createActions(postApi, { ignoreExtraParam: true });
    const reducer = createReducer(actions, {
      defaultData: {},
      dataGetters: {
        [actions.get as any]: dataGetter
      },
      defaultDataGetter
    });

    if (typeof actions.getById === 'function') {
      const result = await actions.getById(10)(jest.fn());

      const rs = reducer({}, actions.getById.success(result.data));
      expect(rs).toMatchSnapshot();
    }
  });
});
