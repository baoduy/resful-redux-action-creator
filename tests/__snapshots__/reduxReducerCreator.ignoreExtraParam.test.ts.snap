// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Reducer Creator Test Create Redux Real Reducer 1`] = `
Object {
  "data": Object {
    "body": "quo et expedita modi cum officia vel magni
doloribus qui repudiandae
vero nisi sit
quos veniam quod sed accusamus veritatis error",
    "id": 10,
    "title": "optio molestias id quia eum",
    "userId": 1,
  },
  "loading": false,
}
`;
